### PR TITLE
Sc 3785/databricks delete file from dbfs

### PR DESCRIPTION
**Pull request by Shamwayne**

### CODE REVIEW

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    from . import helpers
+    from . import errors
+
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--access-token', dest='access_token', required=True)
+    parser.add_argument('--instance-id', dest='instance_id', required=True)
+    parser.add_argument('--source-file-name', dest='source_file_name', required=True)
+    parser.add_argument('--source-folder-name', dest='source_folder_name', required=True)
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py]: Not required. Default to `/FileStore/`

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        print(f"File delete failed. Response: {delete_response.text}")
+        sys.exit(errors.EXIT_CODE_UNKNOWN_ERROR)
+        
+
+def main():
+    args = get_args()
+    access_token = args.access_token
+    instance_id = args.instance_id
+    source_file_name = args.source_file_name
+    source_folder_name = args.source_folder_name
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py:99]: Needs to use shipyard_utils cleaning

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+def delete_file_from_dbfs(client, file_path_and_name):
+    delete_endpoint = "/dbfs/delete"
+    data = {
+        'path': file_path_and_name
+    }
+    print(f"Start delete for {file_path_and_name}")
+    delete_response = client.post(delete_endpoint, data=data)
+    if delete_response.status_code == requests.codes.ok:
+        print(f"DBFS File: {file_path_and_name} delete function completed successfully")
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py]: Returns success message even if the file did not exist. Should catch for that and return separate error.

**Review by blakeburch — CHANGES_REQUESTED** [review]
1) Should support deleting multiple files through regex match.
2) Should have error handling for multiple different situations.

File: databricks_blueprints/delete_file_from_dbfs.py (hunk 2 above)
Comment (blakeburch) [reply databricks_blueprints/delete_file_from_dbfs.py:99]: @shamwayne this still needs to use shipyard_utils for cleaning the folder name.

If I provide the folder name as `/FileStore`, `/FileStore/` or `FileStore/`, it should work in all cases. Right now, if I don't provide the folder with slashes before and after, it returns  Response: {"error_code":"INVALID_PARAMETER_VALUE","message":"Path must be absolute: FileStore/10.csv"}

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        source_folder_name = '/FileStore/'
+
+    # create client
+    client = helpers.DatabricksClient(access_token, instance_id)
+    # get list of all potential file matches
+    if source_file_name_match_type == 'regex_match':
+        files = helpers.list_dbfs_files(client, source_folder_name)
+        matching_file_names = shipyard.files.find_all_file_matches(files,
+                                            re.compile(source_file_name))
+        print(f'{len(matching_file_names)} files found. Preparing to delete...')
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py]: Needs to verify if there are 0 matches and error out. Right now it "successfully deletes 0 files" which isn't the intended behavior.

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            sys.exit(errors.EXIT_CODE_DBFS_SERVICE_UNAVAILABLE)
+    else:
+        print(f"File delete failed. Response: {delete_response.text}")
+        sys.exit(errors.EXIT_CODE_UNKNOWN_ERROR)
+        
+
+def main():
+    args = get_args()
+    access_token = args.access_token
+    instance_id = args.instance_id
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py]: Needs to error handle a bad instance_id correctly.

**Review by blakeburch — CHANGES_REQUESTED** [review]
1) The File is being deleted, but it's always returning the following error:
```
Traceback (most recent call last):
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/delete_file_from_dbfs.py", line 97, in <module>
    main()
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/delete_file_from_dbfs.py", line 93, in main
    delete_file_from_dbfs(client, file_path_and_name)
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/delete_file_from_dbfs.py", line 51, in delete_file_from_dbfs
    check_file_status(client, file_path_and_name)
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/delete_file_from_dbfs.py", line 33, in check_file_status
    if status_code in requests.codes.ok: # File found
TypeError: argument of type 'int' is not iterable
```

This is particularly troublesome when trying to delete multiple files with regex. It instead just deletes the first one and never gets to the rest.
2) I didn't test this (so please do so once it's fixed). Regex match should find all matches across nested subfolders. Make sure if files exist in multiple subfolders on Databricks that it captures and deletes them all appropriately.

File: databricks_blueprints/delete_file_from_dbfs.py (hunk 4 above)
Comment (Shamwayne) [reply databricks_blueprints/delete_file_from_dbfs.py]: Noted, will fix

Comment (Shamwayne) [reply databricks_blueprints/delete_file_from_dbfs.py]: Shipyard Utils currently has clean_folder_name which returns the file stripped of all the leading and trailing backslashes (/). I haven't seen any functionality that deals specifically with creating paths **with** leading and trailing backslashes

Comment (blakeburch) [reply databricks_blueprints/delete_file_from_dbfs.py]: To be clear, we don't want to create paths with leading and trailing backslashes. We just want to format the user input into something that's usable by the API we're working with.

Users don't understand folder formatting concerns. That's our job to abstract it away. If the user provides `FileStore/folder`, it's clear what they want is `/FileStore/folder/`. If you clean the folder using our shipyard_utils functions before submitting it to the API, it should do the trick. If not, something custom might need to be made for Databricks.

**Comment by Shamwayne** [comment]
Please check. Re-requested for a review

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        # create delete file path
+        for file_path_and_name in matching_file_names:
+            delete_file_from_dbfs(client, file_path_and_name)
+    else:
+        file_path_and_name = shipyard.files.combine_folder_and_file_name(source_folder_name,
+                                                                   source_file_name)
+        delete_file_from_dbfs(client, file_path_and_name)
+        
+    
+if __name__ == "__main__":
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py:125]: I think the language needs to be changed to indicate "{file_path_and_name} was successfully deleted"

File: databricks_blueprints/delete_file_from_dbfs.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        # Check if incorrect url provided first
+        if 'nodename nor servname provided' in str(e):
+            print("Invalid or wrong databricks instance url provided. Please check and try again")
+            sys.exit(errors.EXIT_CODE_INVALID_INSTANCE)
+        else:
+            print(f"Error occurred when trying move request: {e}")
+            sys.exit(errors.EXIT_CODE_UNKNOWN_ERROR)
+    if delete_response.status_code == requests.codes.ok:
+        print(f"DBFS File: {file_path_and_name} delete function started...")
+        check_file_status(client, file_path_and_name)
```
Comment (blakeburch) [inline databricks_blueprints/delete_file_from_dbfs.py:78]: Need to handle the situation of if the file _isn't_ deleted what happens?

**Review by blakeburch — COMMENTED** [review]
When there are no matches, I get the following error:

Traceback (most recent call last):
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/delete_file_from_dbfs.py", line 110, in <module>
    main()
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/delete_file_from_dbfs.py", line 92, in main
    files = databricks_client.list_dbfs_files(client, source_folder_name)
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/databricks_client.py", line 139, in list_dbfs_files
    file_list += list_dbfs_files(client, file['path'])
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/databricks_client.py", line 139, in list_dbfs_files
    file_list += list_dbfs_files(client, file['path'])
  File "/Users/blake/Documents/github/databricks-blueprints/databricks_blueprints/databricks_client.py", line 137, in list_dbfs_files
    for file in base_dir['files']:
KeyError: 'files'

**Review by blakeburch — DISMISSED** [review]

**Review by blakeburch — APPROVED** [review]

**Pull request merged**